documentation/onboard-dpg-in-sdkautomation/Integrate-dpg-and-apiview-in-sdk-automation-pipeline.md

```diff
@@ -0,0 +1,31 @@
+# Integrate DPG and ApiView in SDK Automation Pipeline
+This document is targeting for everyone who wants to know all the implementation details of SDK automation pipeline.
+It describes the details of integrating DPG and ApiView in SDK Automation Pipeline.
```

Comment:
Note that the existence of ApiView for review depends on service onboarding DPG on day one.

I am a bit worried that service does not aware of DPG and hence do not even decide to onboard, when they create the swagger and ask for API review.
Later when they request the SDK for their GA, rejecting them is not an option, and hence DPG is still required, even maybe months after the API review (another alternative would be HLC by service themselves).

Reply:
I suppose this depends on when you start counting days. I think that service teams often create PRs in the azure-rest-api-specs repo and then work on them for days, weeks, and sometimes months before the API review is scheduled. At any point in this process the service team can create whatever configuration is needed to produce DPG client libraries so that they can be used in the REST API review.

But more importantly, it's not clear to me that DPG libraries _need_ to be available in the very initial REST API reviews of a new service. Often the initial review for a service's REST API focuses on fundamental issues of resource design and hierarchy, path structure, LRO patterns, and other details purely at the REST API level.  I don't think DPG libraries will be useful for these early reviews. They will be useful later, when issues of naming and client structure are considered -- but that is the point when client teams _have_ to be actively engaged in   producing client libraries.  Note that APEX _requires_ SDKs for a public preview -- well before any GA -- so there is little chance that a service team will get to the point of wanting to GA without already having produced and released at least one version of their client libraries.

Reply:
Hi @mikekistler What we want to do is generating DPG libraries after the REST API review and before the REST API PR is merged. It's not reviewing DPG libraries and REST API at the same time. The advantage is that service team can update the swagger PR directly when there is some comments from Arch Board Review, insteading of creating a new PR to update swagger.

Reply:
Maybe we have different goals then, because "reviewing DPG libraries and REST API at the same time" is _exactly_ what I want to do.  I believe this combined review is the mechanism to identify the changes needed in the REST API PR.

Reply:
Get it. When the DPG is integrated into sdk automation pipeline, I think we also can review DPG libraries and REST API at the same time

Reply:
@mikekistler 
Yes, I agree that the "production ready" quality of the SDK would be likely be required at service public preview, and certainly at their GA.

What I mean is that
1. When service do their first API review, they may not have good idea on who is going to build their SDK (SDK team? Self-serve via DPG? Service team?) and how it is built. (Ted's WIP on join session with SDK arch board and ARM should largely solve it).
2. Even if service decide to build SDK with their own resource (i.e., don't do DPG at all), a generated SDK would still be helpful, as reviewer can see the API in a different (maybe familiar) perspective (I actually prefer traditional SDK with models, instead of DPG on this purpose). I think yours and Wei's latter comment is of the similar intention.

A far-fetched example: I was aware very recently that Purview Scanning had a big problem with their swagger on all api-versions (after sign-off and PR merged).
https://github.com/Azure/azure-rest-api-specs/pull/17382#issuecomment-1134327699
I found this only when I am testing the SDK codegen to generate serialization/de-serialization unit test with mock data (so it requires model and CI run).